src/redirect.rs

```diff
@@ -0,0 +1,125 @@
+use anyhow::{anyhow, Result};
+use reqwest::blocking::Client;
+use reqwest::blocking::{Request, Response};
+use reqwest::header::{
+    HeaderMap, AUTHORIZATION, CONTENT_ENCODING, CONTENT_LENGTH, CONTENT_TYPE, COOKIE, LOCATION,
+    PROXY_AUTHORIZATION, TRANSFER_ENCODING, WWW_AUTHENTICATE,
+};
+use reqwest::{Method, StatusCode, Url};
+
+pub struct RedirectFollower<'a, T>
+where
+    T: FnMut(Response, &mut Request) -> Result<()>,
+{
+    client: &'a Client,
+    max_redirects: usize,
+    callback: Option<T>,
+}
+
+impl<'a, T> RedirectFollower<'a, T>
+where
+    T: FnMut(Response, &mut Request) -> Result<()>,
+{
+    pub fn new(client: &'a Client, max_redirects: usize) -> Self {
+        RedirectFollower {
+            client,
+            max_redirects,
+            callback: None,
+        }
+    }
+
+    pub fn on_redirect(&mut self, callback: T) {
+        self.callback = Some(callback);
+    }
+
+    pub fn execute(&mut self, mut first_request: Request) -> Result<Response> {
+        let mut request = clone_request(&mut first_request);
+        let mut response = self.client.execute(first_request)?;
+        let mut remaining_redirects = self.max_redirects - 1;
+
+        while let Some(mut next_request) = get_next_request(request, &response) {
+            if remaining_redirects > 0 {
+                remaining_redirects -= 1;
+            } else {
+                return Err(anyhow!(
+                    "Too many redirects (--max-redirects={})",
+                    self.max_redirects
+                ));
+            }
+            if let Some(ref mut callback) = self.callback {
+                callback(response, &mut next_request)?;
+            }
+            request = clone_request(&mut next_request);
+            response = self.client.execute(next_request)?;
+        }
+
+        Ok(response)
+    }
+}
+
+fn clone_request(request: &mut Request) -> Request {
+    request
+        .body_mut()
+        .as_mut()
+        .map(|b| b.buffer().expect("unable to buffer request"));
+    request.try_clone().unwrap() // guranteed to not fail if body is already buffered
+}
+
+fn get_next_request(mut request: Request, response: &Response) -> Option<Request> {
+    let get_next_url = |request: &Request| {
+        response
+            .headers()
+            .get(LOCATION)
+            .and_then(|location| location.to_str().ok())
+            .and_then(|location| request.url().join(location).ok())
+    };
+
+    match response.status() {
+        StatusCode::MOVED_PERMANENTLY | StatusCode::FOUND | StatusCode::SEE_OTHER => {
+            let next_url = get_next_url(&request)?;
+            let prev_url = request.url().clone();
+            if is_cross_domain_redirect(&next_url, &prev_url) {
+                remove_sensitive_headers(request.headers_mut());
+            }
+            remove_content_headers(request.headers_mut());
```

Comment:
I wonder if there can be credentials embedded in Url that needs to be removed in cross-domain-redirect 🤔?

Reply:
Credentials seem to be always moved from the URL to headers, so nothing to do here.